integration/airflow/openlineage/airflow/listener.py

```diff
@@ -0,0 +1,141 @@
+import logging
+import threading
+import uuid
+import attr
+
+from typing import TYPE_CHECKING, Optional, Callable
+
+from airflow.listeners import hookimpl
+
+from openlineage.airflow.adapter import OpenLineageAdapter
+from openlineage.airflow.extractors import ExtractorManager
+from openlineage.airflow.utils import DagUtils, get_task_location, get_job_name, get_custom_facets
+
+if TYPE_CHECKING:
+    from airflow.models import TaskInstance, BaseOperator
+    from sqlalchemy.orm import Session
+
+
+@attr.s(frozen=True)
+class RunData:
+    run_id: str = attr.ib()
+    task: "BaseOperator" = attr.ib()
+
+
+class RunDataHolder:
+    """Class that stores run data - run_id and task in-memory. This is needed because Airflow
+    does not always pass all runtime info to on_task_instance_success and
+    on_task_instance_failed that is needed to emit events. This is not a big problem since
+    we're only running on worker - in separate process that is always spawned (or forked) on
```

Comment:
With the details for an active task run being transient, are we concerned about workers restarting?

Reply:
Depends on the failure mode. If we fail inside operator, it will get caught by changing state to `FAILED`. 

If we crash - for example due to OOM, then worker will restart, and another job will start - and change state to `RUNNING` again. If this succeeds, we will capture complete data about this run.

If we wanted to track retries, then we'd need to listen to events on scheduler, because particular worker has no option to get this information. Tracking lifecycle of task instance closer would require additional work on Airflow side.

Reply:
We don't have a `RETRY` run state, though I think we should eventually support tracking job retries. I think capturing the `FAIL` run state and setting the job to `RUNNING` again make sense and is good enough for now.